<a href="https://colab.research.google.com/github/NehaBhatia103/BookRecommendationSystem/blob/main/Prac_BookRecommendationSystem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Book Recommendation System using Collaborative Filter.

## 1.Loading required libraries and Data Set

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
books = pd.read_csv("Books.csv")
users = pd.read_csv("Users.csv")
ratings = pd.read_csv("Ratings.csv")

<ipython-input-2-0f7c25a96956>:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("Books.csv")


In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
# Finding shape of the data frame
print("Shape of users data frame: ",users.shape)
print("Shape of Ratings data frame: ",ratings.shape)
print("Shape of Books data frame: ",books.shape)

Shape of users data frame:  (278858, 3)
Shape of Ratings data frame:  (1149780, 3)
Shape of Books data frame:  (271360, 8)


In [7]:
# Finding Null values
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [8]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [9]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [10]:
# Finding duplicate values
users.duplicated().sum()

0

In [11]:
books.duplicated().sum()

0

##2.Data Visualization

##3.Popularity Based Recommender System

Here to get the most popular book we use the formula of average ratings.We will only consider books having number of ratings greater tha 250.


In [12]:
ratings_with_name=ratings.merge(books,on="ISBN")

In [13]:
ratings_with_name.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [14]:
No_of_ratings=ratings_with_name.groupby("Book-Title").count()["Book-Rating"].reset_index()
No_of_ratings.head()

Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1
3   Ask Lily (Young Women of Faith: Lily Series, ...            1
4   Beyond IBM: Leadership Marketing and Finance ...            1

In [15]:
No_of_ratings=No_of_ratings.rename(columns={"Book-Rating":"No_Of_Ratings"})

In [16]:
No_of_ratings.head()

Book-Title  No_Of_Ratings
0   A Light in the Storm: The Civil War Diary of ...              4
1                              Always Have Popsicles              1
2               Apple Magic (The Collector's series)              1
3   Ask Lily (Young Women of Faith: Lily Series, ...              1
4   Beyond IBM: Leadership Marketing and Finance ...              1

In [17]:
Average_ratings=ratings_with_name.groupby("Book-Title").mean()["Book-Rating"].reset_index()
Average_ratings.head()

<ipython-input-17-6f3ad1847129>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  Average_ratings=ratings_with_name.groupby("Book-Title").mean()["Book-Rating"].reset_index()


Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00
2               Apple Magic (The Collector's series)         0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...         8.00
4   Beyond IBM: Leadership Marketing and Finance ...         0.00

In [18]:
Average_ratings=Average_ratings.rename(columns={"Book-Rating":"Average_Ratings"})
Average_ratings

Book-Title  Average_Ratings
0        A Light in the Storm: The Civil War Diary of ...         2.250000
1                                   Always Have Popsicles         0.000000
2                    Apple Magic (The Collector's series)         0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...         8.000000
4        Beyond IBM: Leadership Marketing and Finance ...         0.000000
...                                                   ...              ...
241066                                      Ã?Â?lpiraten.         0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.         5.250000
241068                                Ã?Â?sterlich leben.         7.000000
241069                              Ã?Â?stlich der Berge.         2.666667
241070                                  Ã?Â?thique en toc         4.000000

[241071 rows x 2 columns]

In [19]:
Popularity_DF=No_of_ratings.merge(Average_ratings,on="Book-Title")
Popularity_DF.head()

Book-Title  No_Of_Ratings  \
0   A Light in the Storm: The Civil War Diary of ...              4   
1                              Always Have Popsicles              1   
2               Apple Magic (The Collector's series)              1   
3   Ask Lily (Young Women of Faith: Lily Series, ...              1   
4   Beyond IBM: Leadership Marketing and Finance ...              1   

   Average_Ratings  
0             2.25  
1             0.00  
2             0.00  
3             8.00  
4             0.00

In [20]:
Popularity_DF=Popularity_DF[Popularity_DF["No_Of_Ratings"]>=250].sort_values("Average_Ratings",ascending=False).head(50)

In [21]:
Popularity_DF

Book-Title  No_Of_Ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)            428   
80422        Harry Potter and the Goblet of Fire (Book 4)            387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)            278   
80426   Harry Potter and the Order of the Phoenix (Boo...            347   
80414    Harry Potter and the Chamber of Secrets (Book 2)            556   
191612  The Hobbit : The Enchanting Prelude to The Lor...            281   
187377  The Fellowship of the Ring (The Lord of the Ri...            368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...            575   
211384     The Two Towers (The Lord of the Rings, Part 2)            260   
219741                              To Kill a Mockingbird            510   
183573                                  The Da Vinci Code            898   
187880                 The Five People You Meet in Heaven            430   
180556                             The Catcher in the Rye            449   
196326                          The Lovely Bones: A Novel           1295   
764                                                  1984            284   
144165                           Prodigal Summer: A Novel            253   
128670                                         Neverwhere            265   
206502                            The Secret Life of Bees            774   
168719  Stupid White Men ...and Other Sorry Excuses fo...            283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...            493   
204387                The Red Tent (Bestselling Backlist)            723   
191589               The Hitchhiker's Guide to the Galaxy            268   
129379   Nickel and Dimed: On (Not) Getting By in America            335   
93381                                       Into the Wild            252   
63867                                      Fahrenheit 451            409   
74750                           Girl with a Pearl Earring            526   
136145                                          Outlander            283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...            585   
156102                     Seabiscuit: An American Legend            275   
107962                                         Life of Pi            664   
176845                                     The Bean Trees            389   
2281    A Child Called \It\": One Child's Courage to S...            265   
8434                                       ANGELA'S ASHES            279   
76343                                         Good in Bed            490   
64931   Fast Food Nation: The Dark Side of the All-Ame...            321   
12700                                       American Gods            302   
161645                                 Skipping Christmas            322   
105777  Left Behind: A Novel of the Earth's Last Days ...            318   
189551    The Golden Compass (His Dark Materials, Book 1)            336   
181679                                   The Color Purple            314   
160336                               Silence of the Lambs            256   
8752                                          About a Boy            262   
158138                  Seven Up (A Stephanie Plum Novel)            278   
175097  The Alchemist: A Fable About Following Your Dream            266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...            269   
170101                       Suzanne's Diary for Nicholas            457   
111073                                  Lord of the Flies            259   
5664                              A Prayer for Owen Meany            413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)            301   
233851       White Oleander : A Novel (Oprah's Book Club)            356   

        Average_Ratings  
80434          5.852804  
80422          5.824289  
80441          5.737410  
80426          5.501441  
80414          5.183453  
191612        

In [22]:
Popularity_DF=Popularity_DF.merge(books,on="Book-Title").drop_duplicates("Book-Title")[["Book-Title","Book-Author","Image-URL-M","No_Of_Ratings","Average_Ratings"]]

In [23]:
Popularity_DF

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            George Orwell   
72       Barbara Kingsolver   
73              Neil Gaiman   
78            Sue Monk Kidd   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


##3.Collaborative Filtering Based Recommender System
Here we will be taking books which are rated by more than 200 users and those users who have rated more than 50 books.

In [24]:
ratings_with_name.groupby("User-ID").count()["Book-Rating"]

User-ID
2          1
8         17
9          3
10         1
12         1
          ..
278846     1
278849     4
278851    23
278852     1
278854     8
Name: Book-Rating, Length: 92106, dtype: int64

out of total users only 92106 users have rated any book.

In [25]:
X = ratings_with_name.groupby("User-ID").count()["Book-Rating"]>200
X[X]

User-ID
254       True
2276      True
2766      True
2977      True
3363      True
          ... 
274308    True
275970    True
277427    True
277639    True
278418    True
Name: Book-Rating, Length: 811, dtype: bool

In [26]:
Frequent_Readers = X[X].index
Frequent_Readers

Int64Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,
              6323,   6543,
            ...
            271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427,
            277639, 278418],
           dtype='int64', name='User-ID', length=811)

In [27]:
Filtered_Ratings = ratings_with_name[ratings_with_name["User-ID"].isin(Frequent_Readers)]
Filtered_Ratings

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [28]:
Popularity_DF["Image-URL-M"][0]


'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'

out of total users only 474007 users were Frequent users who have rated the books

In [29]:
y = Filtered_Ratings.groupby("Book-Title").count()["Book-Rating"]>=50
Famous_Books = y[y].index

In [30]:
Final_Ratings = Filtered_Ratings[Filtered_Ratings["Book-Title"].isin(Famous_Books)]

we have 58586 famous books

In [31]:
# Pivot TAble

Pivot_table = Final_Ratings.pivot_table(index="Book-Title",columns="User-ID",values="Book-Rating")

In [32]:
Pivot_table.fillna(0,inplace=True)
Pivot_table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

we can imagine each book as a point in 810 dimentional system and we will find the eucledian distance to find the similar books.

In [33]:
from sklearn.metrics.pairwise import cosine_similarity

In [34]:
Similarity_Score = cosine_similarity(Pivot_table)

In [35]:
Similarity_Score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

first array is discribing the distance of the first book(1984) with other books

In [36]:
list(enumerate(Similarity_Score[0]))  # it gives the similiarity score of the book 1984 with all other books


[(0, 0.9999999999999999),
 (1, 0.10255024829874465),
 (2, 0.012208555646509366),
 (3, 0.0),
 (4, 0.053672244225442485),
 (5, 0.027749013024662858),
 (6, 0.08216491130000354),
 (7, 0.13732869060263378),
 (8, 0.03261686350494277),
 (9, 0.03667590746512177),
 (10, 0.02322418425912186),
 (11, 0.06766486753662343),
 (12, 0.020839782065795794),
 (13, 0.09673734976610586),
 (14, 0.13388864701187034),
 (15, 0.08303111691811028),
 (16, 0.1115354275858341),
 (17, 0.05100410510067514),
 (18, 0.025177841632809915),
 (19, 0.11706383185253948),
 (20, 0.0),
 (21, 0.14333792815336935),
 (22, 0.07847534327232125),
 (23, 0.06150451339525138),
 (24, 0.08723968326958673),
 (25, 0.0),
 (26, 0.07009814362931202),
 (27, 0.13658680855472938),
 (28, 0.07600327535501016),
 (29, 0.12167133598776536),
 (30, 0.007680459111722776),
 (31, 0.01473221255132234),
 (32, 0.0),
 (33, 0.07965814289647455),
 (34, 0.04522616573040869),
 (35, 0.01556270914112474),
 (36, 0.09495937740802216),
 (37, 0.018230696811058135),
 (38,

In [37]:
# to get top 5  sorted according to the similarity in ascending way
sorted(list(enumerate(Similarity_Score[0])),key=lambda x:x[1],reverse=True)[1:6]

[(47, 0.2702651417103732),
 (545, 0.2639619371123496),
 (82, 0.2366937434740099),
 (634, 0.23299389358170397),
 (551, 0.2262639743141286)]

In [38]:
## creating the function
def recommend(book_name):
  index = np.where(Pivot_table.index==book_name)[0][0]
  similar_items = sorted(list(enumerate(Similarity_Score[index])),key=lambda x:x[1],reverse=True)[1:6]

  for i in similar_items:
    print(Pivot_table.index[i[0]])

In [39]:
# Test it for roamntic Novels
recommend("Message in a Bottle")

Nights in Rodanthe
The Mulberry Tree
A Walk to Remember
River's End
Nightmares &amp; Dreamscapes


In [ ]:
recommend("The Notebook")

A Walk to Remember
The Red Tent (Bestselling Backlist)
The Reader
Suzanne's Diary for Nicholas
The Prince of Tides


In [40]:
import pickle
pickle.dump(Popularity_DF,open("popular.pkl","wb"))

In [ ]:
books.drop_duplicates('Book-Title')

In [44]:
pickle.dump(Pivot_table,open('pt.pkl','wb'))
pickle.dump(books,open('books.pkl','wb'))
pickle.dump(Similarity_Score,open('similarity_scores.pkl','wb'))